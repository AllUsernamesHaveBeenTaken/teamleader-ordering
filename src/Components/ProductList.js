import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import ProductItem from './ProductItem';
import AddProduct from './AddProduct';
import { Modal } from '../Elements';
import { Toggle } from '../Utilities';

const ProductList = ({ items, orderId, success }) => (
  <div>
    <h2>Products in this order</h2>
    <table style={{ borderCollapse: 'collapse' }}>
      <thead>
        <tr>
          <th>Product id</th>
          <th>Quantity</th>
          <th>Unit price</th>
          <th>Total</th>
        </tr>
      </thead>
      <tbody>
        {items.map(item => (
          <ProductItem
            key={`item-${item['product-id']}`}
            success={success}
            item={item}
            orderId={orderId}
          />
        ))}
        <tr>
          <th align="left">
            <Toggle>
              {({ on, toggle }) => (
                <Fragment>
                  <button type="button" onClick={toggle}>
                    +
                  </button>
                  <Modal on={on} toggle={toggle}>
                    <AddProduct />
                  </Modal>
                </Fragment>
              )}
            </Toggle>
          </th>
        </tr>
      </tbody>
    </table>
  </div>
);

export default ProductList;

ProductList.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      'product-id': PropTypes.string.isRequired,
      quantity: PropTypes.string.isRequired,
      'unit-price': PropTypes.string.isRequired,
      total: PropTypes.string.isRequired,
    }),
  ).isRequired,
  orderId: PropTypes.string.isRequired,
  success: PropTypes.bool,
};

ProductList.defaultProps = {
  success: false,
};
