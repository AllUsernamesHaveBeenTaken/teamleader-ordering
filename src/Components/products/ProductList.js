import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import { ProductContext } from '../../Contexts';
import ProductItem from './ProductItem';
import AddProduct from './AddProduct';
import { Modal } from '../../Elements';
import { Toggle } from '../../Utilities';

const ProductList = ({
  items, orderId, success, addNewItem,
}) => (
  <div>
    <h2>Products in this order</h2>
    <table style={{ borderCollapse: 'collapse' }}>
      <thead>
        <tr>
          <th>Product id</th>
          <th>Quantity</th>
          <th>Unit price</th>
          <th>Total</th>
        </tr>
      </thead>
      <tbody>
        {items.map(item => (
          <ProductItem
            key={`item-${item['product-id']}`}
            success={success}
            item={item}
            orderId={orderId}
          />
        ))}
        {!success && (
          <tr>
            <th align="left">
              <Toggle>
                {({ on, toggle }) => (
                  <Fragment>
                    <button type="button" onClick={toggle}>
                      +
                    </button>
                    <Modal on={on} toggle={toggle}>
                      <ProductContext.Consumer>
                        {({ products }) => (
                          <AddProduct
                            orderId={orderId}
                            products={products}
                            addNewItem={addNewItem}
                          />
                        )}
                      </ProductContext.Consumer>
                    </Modal>
                  </Fragment>
                )}
              </Toggle>
            </th>
          </tr>
        )}
      </tbody>
    </table>
  </div>
);

export default ProductList;

ProductList.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      'product-id': PropTypes.string.isRequired,
      quantity: PropTypes.string.isRequired,
      'unit-price': PropTypes.string.isRequired,
      total: PropTypes.string.isRequired,
    }),
  ).isRequired,
  orderId: PropTypes.string.isRequired,
  success: PropTypes.bool,
  addNewItem: PropTypes.func.isRequired,
};

ProductList.defaultProps = {
  success: false,
};
