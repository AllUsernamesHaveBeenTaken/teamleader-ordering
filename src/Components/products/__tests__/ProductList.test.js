import React from 'react';
import { render, cleanup } from 'react-testing-library';
import ProductList from '../ProductList';

afterEach(() => {
  cleanup();
  console.error.mockClear();
});

console.error = jest.fn();

const substractItem = jest.fn();
const addItem = jest.fn();
const deleteItem = jest.fn();
const addNewItem = jest.fn();

const orderId = '1';

const items = [
  {
    'product-id': 'B102',
    quantity: '10',
    total: '49.90',
    'unit-price': '4.99',
  },
  {
    'product-id': 'B102',
    quantity: '10',
    total: '49.90',
    'unit-price': '4.99',
  },
];
// ProductList's props = items, orderId, success, addNewItem, addItem, deleteItem, substractItem,
test('<ProductList /> where success is true', () => {
  const success = true;
  const { getByTestId, queryByTestId, getAllByTestId } = render(
    <ProductList
      items={items}
      orderId={orderId}
      success={success}
      addNewItem={addNewItem}
      substractItem={substractItem}
      addItem={addItem}
      deleteItem={deleteItem}
    />,
  );

  expect(getByTestId('items-table')).toBeTruthy();

  expect(queryByTestId('add-product')).toBeFalsy();
  expect(addNewItem).toHaveBeenCalledTimes(0);

  expect(getAllByTestId('item-row').length).toBe(items.length);
});
