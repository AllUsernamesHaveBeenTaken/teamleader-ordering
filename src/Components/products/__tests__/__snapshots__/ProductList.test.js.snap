// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductList /> where success is true 1`] = `
<div>
  <h2>
    Products in this order
  </h2>
  <table
    data-testid="items-table"
    style="border-collapse: collapse;"
  >
    <thead>
      <tr>
        <th>
          Product id
        </th>
        <th>
          Quantity
        </th>
        <th>
          Unit price
        </th>
        <th>
          Total
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-bZQynM bGQAfj"
        data-testid="item-row"
      >
        <td>
          B102
        </td>
        <td>
          10
        </td>
        <td>
          4.99
        </td>
        <td>
          49.90
        </td>
      </tr>
      <tr
        class="sc-bZQynM bGQAfj"
        data-testid="item-row"
      >
        <td>
          B102
        </td>
        <td>
          10
        </td>
        <td>
          4.99
        </td>
        <td>
          49.90
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
